ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,481 @@
+ *       b      loop
+ *   end:
+ *       ...
+ *
+ * TODO i#5036 Add support for scatter store operations.
+ */
+static void
+expand_scalar_plus_vector(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
+                          const scatter_gather_info_t *sg_info, reg_id_t scratch_gpr,
+                          reg_id_t scratch_pred, app_pc orig_app_pc)
+{
+#define EMIT(op, ...)    \
+    instrlist_preinsert( \
+        bb, sg_instr, INSTR_XL8(INSTR_CREATE_##op(drcontext, __VA_ARGS__), orig_app_pc))
+
+    DR_ASSERT_MSG(reg_is_z(sg_info->index_reg), "Index must be a Z register");
+
+    if (sg_info->is_load) {
+        /* First we deal with the inactive elements. Gather loads are always zeroing so we
+         * need to set all inactive elements to 0.
+         */
+        if (sg_info->index_reg == sg_info->gather_dst_reg) {
+            /* The dst register is also the index register so we need to preserve the
+             * value of the active elements so we can use them as offsets. We do this by
+             * cpying a 0 value into the dst register using the inverse of the mask_reg as
+             * the governing predicate.
+             */
+
+            /* ptrue    scratch_pred.b */
+            EMIT(ptrue_sve, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
+                 opnd_create_immed_pred_constr(DR_PRED_CONSTR_ALL));
+
+            /* not      scratch_pred.b, scratch_pred/z, mask_reg.b */
+            EMIT(eor_sve_pred_b, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
+                 opnd_create_predicate_reg(scratch_pred, false),
+                 opnd_create_reg_element_vector(sg_info->mask_reg, OPSZ_1),
+                 opnd_create_reg_element_vector(scratch_pred, OPSZ_1));
+
+            /* cpy      gather_dst_reg.elsz, scratch_pred/m, #0, lsl #0 */
+            EMIT(cpy_sve_shift_pred,
+                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
+                                                sg_info->element_size),
+                 opnd_create_predicate_reg(scratch_pred, true), OPND_CREATE_INT8(0),
+                 opnd_create_immed_uint(0, OPSZ_1b));
+        } else {
+            /* We don't care about any values in the dst register so zero the whole thing.
+             */
+
+            /* dup      gather_dst_reg.elsz, #0, lsl #0 */
+            EMIT(dup_sve_shift,
+                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
+                                                sg_info->element_size),
+                 OPND_CREATE_INT8(0), opnd_create_immed_uint(0, OPSZ_1b));
+        }
+    }
+
+    /* pfalse   scratch_pred.b */
+    EMIT(pfalse_sve, opnd_create_reg_element_vector(scratch_pred, OPSZ_1));
+
+    instr_t *loop_label = INSTR_CREATE_label(drcontext);
+    instr_t *end_label = INSTR_CREATE_label(drcontext);
+
+    instrlist_meta_preinsert(bb, sg_instr, loop_label);
+
+    /* pnext scratch_pred.elsz, mask_reg, scratch_pred.elsz */
```

Comment:
What does `elsz` mean?

Reply:
Its short for element size. The abbreviation is used in the codec but its a bit cryptic if you aren't familiar. I'll change it to `element_size`.

Reply:
It depends on the instruction. Instructions (like pnext) which treat the register value as a mask generally need to know the element size, but bitwise operations (like pfalse, and, orr, etc) just operate on the register value and are conventionally written with the element size `b` to indicate it operates on every bit in the register.